heat/core/dndarray.py

```diff
@@ -1563,54 +1563,64 @@ def __getitem__(self, key):
+                    gout_full.append(slice_size)
+                slice_size = 0
+            elif isinstance(k, int):
+                slice_size = 1
+            elif isinstance(k, (list, DNDarray, torch.Tensor, np.ndarray)):
+                if not advanced_ind:
+                    # first dimension with advanced indexing: keep
+                    if isinstance(k, list):
+                        slice_size = len(k)
+                    else:
+                        slice_size = k.shape[0] if not kgshape_flag else kgshape[c]
+                    advanced_ind = True
+                else:
+                    # drop dimension, correct new_split if necessary
+                    slice_size = 0
+                    if self.split is not None and c <= self.split:
+                        new_split -= 1
+            if slice_size > 1:
+                gout_full.append(slice_size)
+            elif slice_size == 1 and self.split is not None:
+                if c == self.split:
+                    gout_full.append(slice_size)
+                elif c < self.split:
+                    new_split -= 1
             if isinstance(k, DNDarray):
                 key[c] = k.larray
-        if all(g == 1 for g in gout_full):
-            gout_full = [1]
-        else:
-            # delete the dimensions from gout_full if they are not touched (they will be removed)
-            for i in range(len(gout_full) - 1, -1, -1):
-                if gout_full[i] is None:
-                    del gout_full[i]
+
+        # finally, add back unsliced dimensions if any
+        left_overs = self.ndim - len(key)
+        if left_overs > 0:
+            gout_full = gout_full + list(self.gshape[-left_overs:])
 
         key = tuple(key)
         if not self.is_distributed():
-            if not self.comm.size == 1:
-                return factories.array(
-                    self.__array[key],
-                    dtype=self.dtype,
-                    split=self.split,
-                    device=self.device,
-                    comm=self.comm,
-                )
-            else:
-                gout = tuple(self.__array[key].shape)
-                if self.split is not None and self.split >= len(gout):
-                    new_split = len(gout) - 1 if len(gout) - 1 >= 0 else None
-                else:
-                    new_split = self.split
-
-                return DNDarray(
-                    self.__array[key],
-                    gout,
-                    self.dtype,
-                    new_split,
-                    self.device,
-                    self.comm,
-                    self.balanced,
-                )
+            arr = self.__array[key].reshape(gout_full)
+            return DNDarray(
+                arr, tuple(gout_full), self.dtype, new_split, self.device, self.comm, self.balanced
+            )
+
         # else: (DNDarray is distributed)
```

Comment:
Because the `comm.chunk` is used, the DNDarray has to be distributed and balanced, right? 
For distributed and unbalanced, the `lshape_map`  should be used instead of `comm.chunk`

Reply:
Yes, this problem is the reason why we implemented the `balanced` property, see #668 and #681. I've now implemented a `DNDarray.counts_displs` method to provide the actual counts and offsets for the DNDarray, which in turn uses `create_lshape_map`, which I have modified to only need communication when the DNDarray is not balanced.